### DISCUSSION TITLE
how to get Object position in group?

### DISCUSSION THREAD

**Discussion by vaynevayne** [fabricjs/fabric.js#10303] [Q&A] [answered]
I'm not familiar with this method. The coordinates returned by this method are the same as canva.getScenePoint (e), right?
Previously I used the following method to calculate. Want to ask why it's better? Like take rect.aCoords directly or something like that? Or is there a way to get it directly?

```jsx
const position = fabric.util.transformPoint(
  // you can choose point of object (left/center/right, top/center/bottom)
  object.getPointByOrigin('left', 'top'), 
  object.calcTransformMatrix()
)
```
Also please aCoords oCoord What does that mean? Can I use it directly here?

**Comment by vaynevayne** [comment]
@zhe-he  大佬有空科普下吗

**Comment by vaynevayne** [comment]
https://github.com/fabricjs/fabric.js/blob/992e003f659d2801e901a2d9381dc1fd4b1c3a53/src/shapes/Object/ObjectGeometry.ts#L204
看了源码 好像getCoords 就行

**Comment by zhe-he** [comment, marked as answer]
aCoords represents the position of the object relative to its parent.
getCoords retrieves the position of the object relative to the canvas.
oCoords refers to the control points of the object when it is selected, relative to the canvas.

```ts
// 把每个图形看成一个点， 如果这个点相对画布位置(10，10)，那么我们记为translateX = 10, translateY = 10
// 如果一个点没有尺寸，它是无法进行缩放和旋转的，那么我们可以把这个点的宽度和高度分别记为单位长度1，originX和originY应运而生
// 对这个图形进行旋转30度时， 我们可以记为 rotate(30deg)
// 对这个图形进行缩放2倍时，我们可以记为 scale(2)
// 把上述3个变换我们合并称为 ownMatrix -> this.calcOwnMatrix()

// 如果把一个变换的图形A放入另一个变换的组G内
// 当A应用G的变换后，才是我们看到的样子，我们记为 matrix = GMatrix * AMatrix   -> this.calcTransformMatrix()
// 所以矩阵计算它更好
```

**Reply by vaynevayne** [reply]
aCoords 和 oCoords 是不是说反了?

**Reply by zhe-he** [reply]
<img width="441" alt="image" src="https://github.com/user-attachments/assets/f89fe5c5-8fe3-46eb-b606-1b71692c8936">

Does this image look a bit clearer to you?

**Reply by vaynevayne** [reply]
有具体的场景就好理解了
aCoords 是世界坐标系, 也就是absolute, 但是有时候不更新, 需要使用getCoords 获取一下,
oCoords 没用过 暂时不知道这个干嘛的

**Reply by zhe-he** [reply]
aCoords 是相对父亲的坐标
如果父亲是组，是相对组的坐标，如果没有父亲，是相对画布的坐标。
getCoords 返回的永远是相对画布的坐标。
aCoords 有缓存，当元素通过 this.set 来设置 left/top/skewX/skewY/scaleX/scaleY/angle时，aCoords不会更新，需要调用this.setCoords() 来更新缓存。

oCoords 是 控制点 距离 dom元素canvas 左上角的距离，它与画布缩放平移无关，只与画布的尺寸相关。
比如 你可以理解为 在 一个 div中 画了一个点，这个点距离div左上角顶点的距离，相当于css样式中距离。

**Reply by asturur** [reply]
We refer to scenePoint as a point that is in the design/composition/drawing space. All scene points are equal.
aCoords and oCoords are really just a tool for caching and calculations.
You shouldn't work with oCoords unless you are doing something specific with controls, but you should be covered by the api that fabricJS uses for targeting controls already. You can work with aCoords once you understand those are mostly for targeting and dimensions calculations in groups,  and while they match for the Rect object with its' size, the similarities basically ends there.
aCoords are relative to the parent and are also stale data like oCoords.
They are good values when fabricJS sets them but once you change left, top, scale or other properties those values are not useful anymore.

**Comment by vaynevayne** [comment]
谢谢
thanks
I want to know the location of a rectangle in the browser to locate its dom control. How should I calculate it? Consider zoom checkpoint

**Reply by asturur** [reply]
let's assume your aCoords are good, you can do 2 things:

object.getCoords() you have scene points, now from scenePoints to html points you have to use the sendPointToPlane function:
```
htmlPoint = sendPointToPlane(point, canvas.viewportTransform, undefined)
```

Maybe this is a good list of explanations to add to the website

Once you have the html points, the canvas has its top/left position on the dom that you can get with ``` canvas._offset ```

**Comment by vaynevayne** [comment]
thanks, very good
